mysql/mydecimal.go

```diff
@@ -0,0 +1,776 @@
+		carry = 1
+		sum -= wordBase
+	} else {
+		carry = 0
+	}
+	return sum, carry
+}
+
+// fixWordCntError limits word count in wordBufLen, and returns overflow or truncate error
+func fixWordCntError(wordsInt, wordsFrac int) (newWordsInt, newWordsFrac, errcode int) {
+	if wordsInt+wordsFrac > wordBufLen {
+		if wordsInt > wordBufLen {
+			return wordBufLen, 0, eDecOverflow
+		}
+		return wordsInt, wordBufLen - wordsInt, eDecTruncate
+	}
+	return wordsInt, wordsFrac, eDecOK
+}
+
+/*
+  countLeadingZeroes returns the number of leading zeroes that can be removed from fraction.
+
+  @param   i    start index
+  @param   word value to compare against list of powers of 10
+*/
+func countLeadingZeroes(i int, word int32) int {
+	leading := 0
+	for word < powers10[i] {
+		i--
+		leading++
+	}
+	return leading
+}
+
+/*
+  countTrailingZeros returns the number of trailing zeroes that can be removed from fraction.
+
+  @param   i    start index
+  @param   word  value to compare against list of powers of 10
+*/
+func countTrailingZeroes(i int, word int32) int {
+	trailing := 0
+	for word%powers10[i] == 0 {
+		i++
+		trailing++
+	}
+	return trailing
+}
+
+// MyDecimal represents a decimal value.
+type MyDecimal struct {
+	// the number of *decimal* digits before the point.
+	digitsInt int
+
+	// the number of decimal digits after the point
+	digitsFrac int
+
+	negative bool
+
+	// an array of int32 words.
+	// A word is an int32 value can hold 9 digits.
+	// 0 <= word < wordBase
+	wordBuf [maxWordBufLen]int32
+}
+
+func (d *MyDecimal) String() string {
+	return fmt.Sprintf("di:%d, df:%d, neg:%v, buf:%v", d.digitsInt, d.digitsFrac, d.negative, d.wordBuf)
+}
+
+// FromString parse decimal from string.
+func (d *MyDecimal) FromString(str []byte) int {
+	ec := eDecBadNum
+	for i := 0; i < len(str); i++ {
+		if !isSpace(str[i]) {
+			str = str[i:]
+			break
+		}
+	}
+	if len(str) == 0 {
+		*d = zeroMyDecimal
+		return eDecBadNum
+	}
+	switch str[0] {
+	case '-':
+		d.negative = true
+		fallthrough
+	case '+':
+		str = str[1:]
+	}
+	var strIdx int
+	for strIdx < len(str) && isDigit(str[strIdx]) {
+		strIdx++
+	}
+	digitsInt := strIdx
+	var digitsFrac int
+	var endIdx int
+	if strIdx < len(str) && str[strIdx] == '.' {
+		endIdx = strIdx + 1
+		for endIdx < len(str) && isDigit(str[endIdx]) {
+			endIdx++
+		}
+		digitsFrac = endIdx - strIdx - 1
+	} else {
+		digitsFrac = 0
+		endIdx = strIdx
+	}
+	if digitsInt+digitsFrac == 0 {
+		*d = zeroMyDecimal
+		return eDecBadNum
+	}
+	wordsInt := (digitsInt + digitsPerWord - 1) / digitsPerWord
+	wordsFrac := (digitsFrac + digitsPerWord - 1) / digitsPerWord
+	wordsInt, wordsFrac, ec = fixWordCntError(wordsInt, wordsFrac)
+	if ec != eDecOK {
+		digitsFrac = wordsFrac * digitsPerWord
+		if ec == eDecOverflow {
+			digitsInt = wordsInt * digitsPerWord
+		}
+	}
+	d.digitsInt = digitsInt
+	d.digitsFrac = digitsFrac
+	wordIdx := wordsInt
+	strIdxTmp := strIdx
+	var word int32
+	var innerIdx int
+	for digitsInt > 0 {
```

Comment:
this code patten is redundancy with line 214~228, extract to a function maybe better?